modyn/supervisor/internal/pipeline_executor/evaluation_executor.py

```diff
@@ -295,30 +329,52 @@ def _single_batched_evaluation(
-                f"{interval_end} not started with reason: {failure_reason}."
-            )
-            return failure_reason, {}
-
-        logger.info(f"Evaluation started for model {model_id_to_eval} on split {interval_start} " f"to {interval_end}.")
+            failure_reasons: list[tuple[str | None, dict]] = []
+            # note: interval indexes correspond to the intervals in the request
+            for interval_idx, interval_response in enumerate(response.interval_responses):
+                if interval_response.eval_aborted_reason != EvaluationAbortedReason.NOT_ABORTED:
+                    reason = get_failure_reason(interval_response.eval_aborted_reason)
+                    failure_reasons.append((reason, {}))
+                    logger.error(
+                        f"Evaluation for model {model_id_to_eval} on split {intervals[interval_idx]} "
+                        f"not started with reason: {reason}."
+                    )
+            return failure_reasons
+
+        logger.info(f"Evaluation started for model {model_id_to_eval} on intervals {intervals}.")
         self.grpc.wait_for_evaluation_completion(response.evaluation_id)
         eval_data = self.grpc.get_evaluation_results(response.evaluation_id)
         self.grpc.cleanup_evaluations([response.evaluation_id])
-        # here we assume only one interval is evaluated
-        eval_results: dict = {"dataset_size": response.interval_responses[0].dataset_size, "metrics": []}
 
-        assert len(eval_data) == 1, "The evaluation request only contains one interval so we expect only one result."
-        assert eval_data[0].interval_index == 0
-        for metric in eval_data[0].evaluation_data:
-            eval_results["metrics"].append({"name": metric.metric, "result": metric.result})
-
-        return None, eval_results
+        eval_results: list[tuple[str | None, dict[str, Any]]] = []
+        assert len(eval_data) == len(intervals), f"We expected {len(intervals)} intervals, but got {len(eval_data)}."
+        assert eval_data
+
+        # as the dataset_size comes from the `EvaluateModelResponse` and response is a dense list (no intervals are
+        # skipped), we build a list of results with the same order as the intervals.
+        # the metrics will be filled in the next loop that unwraps `EvaluationResultResponse`.
+        for interval_response in response.interval_responses:
+            if interval_response.eval_aborted_reason != EvaluationAbortedReason.NOT_ABORTED:
+                reason = get_failure_reason(interval_response.eval_aborted_reason)
+                eval_results.append((reason, {}))
+            else:
+                eval_results.append((None, {"dataset_size": interval_response.dataset_size, "metrics": []}))
+
+        for interval_result in eval_data:
+            interval_idx = interval_result.interval_index
+            assert eval_results[interval_idx][0] is None, "Evaluation failed, no metrics should be present."
```

Comment:
I'm sure there is a logic here but I don't understand. We are asserting for every interval that the failure reason is None. But what about the failed evaluations? Then this assertion will fail, won't it? Should we really crash then, instead of just not logging any metrics like before? Also, I don't fully understand the assertion message here

Reply:
actually in the function that calls this (the worker func) we handle failure reasons but i am not sure about this assertion and what its purpose is

Reply:
I assumed that in the `eval_data` response, unsuccessful evaluations wouldn't show up.

see this comment in the evaluator.proto

```

  // Since not every interval evaluation from EvaluateModelRequest may be successful,
  // the EvaluationIntervalData contained in the EvaluationResultResponse must explicitly specify what interval this
  // evaluation data corresponds to. The interval_index is the index of the interval in the list
  // Datainfo.evaluation_intervals in the EvaluateModelRequest.
  // For example if Datainfo.evaluation_intervals have 3 intervals, [interval1, interval2, interval3],
  // and interval2 fails. Then the EvaluationResultResponse will have 2 EvaluationIntervalData, one with interval_index
  // 0 (which corresponds to interval1) and the other with interval_index 2 (which corresponds to interval3).
```

if that assumption does not hold, we will find out soon enough. Better to have one assertion too much than missing oneone😅

Reply:
Ahh, sorry, I got confused (as expected) over the index stuff